service/storage_service.cc

```diff
@@ -1608,6 +1608,50 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
     assert(_group0);
     co_await _group0->finish_setup_after_join(*this, _qp, _migration_manager.local(), _raft_experimental_topology);
     co_await _cdc_gens.local().after_join(std::move(cdc_gen_id));
+
+    if (_raft_experimental_topology) {
+        auto holder = _async_gate.hold();
+        // Waited on during stop()
+        (void)([] (storage_service& me, sharded<db::system_distributed_keyspace>& sys_dist_ks, gate::holder holder) -> future<> {
+            try {
+                co_await me.track_upgrade_progress_to_topology_coordinator(sys_dist_ks);
+            } catch (const abort_requested_exception&) {
+                // Ignore
+            }
+        })(*this, sys_dist_ks, std::move(holder));
+    }
+}
+
+future<> storage_service::track_upgrade_progress_to_topology_coordinator(sharded<db::system_distributed_keyspace>& sys_dist_ks) {
+    _abort_source.check();
+    auto cv_bumper = _abort_source.subscribe([this] () noexcept { _topology_state_machine.event.broadcast(); });
+
+    assert(_group0);
+    co_await _group0->client().wait_until_group0_upgraded(_abort_source);
+
+    // First, wait for the feature to become enabled
+    shared_promise<> p;
+    _feature_service.supports_consistent_topology_changes.when_enabled([&] () noexcept { p.set_value(); });
+    co_await p.get_shared_future(_abort_source);
+    slogger.info("raft topology: The cluster is ready to start upgrade to the raft topology. The procedure needs to be manually triggered. Refer to the documentation");
+
+    // Wait until upgrade is started
+    co_await _topology_state_machine.event.when([this] {
+        return !legacy_topology_change_enabled() || _abort_source.abort_requested();
+    });
+    _abort_source.check();
+    slogger.info("raft topology: upgrade to raft topology has started");
+
+    // Start the topology coordinator monitor fiber. If we are the leader, this will start
+    // the topology coordinator which is responsible for driving the upgrade process.
+    _raft_state_monitor = raft_state_monitor_fiber(_group0->group0_server(), sys_dist_ks);
```

Comment:
What happens if the fiber throws before it gets here?

Reply:
The only exception that can be thrown is `abort_requested_exception`  when waiting on any of the conditions is aborted due to the node shutting down - this was intentional. Any other exception is unexpected.

I can add a `try..catch` and log a message if an unexpected exception occurs.

Reply:
> The only exception that can be thrown is abort_requested_exception

Are you sure? It is a lot of code there.

Reply:
OK, I am not 100% sure. However, the code in `track_upgrade_progress_to_topology_coordinator` before this line just waits for some conditions to happen; the waits are abortable and can throw `abort_requested_exception`. It's not a part of their interface to throw other exceptions.

Of course, there is a risk that some of them may throw `std::bad_alloc` or something similarly unexpected (`when_enabled()` uses `make_lw_shared` so it technically could). Do you suggest to handle them in some other way than printing the error?

Reply:
It is better to either restart the fiber or kill the Scylla since without the fiber it will not work properly.

Reply:
>  It is better to either restart the fiber or kill the Scylla since without the fiber it will not work properly.


Kill Scylla? We decided not to kill Scylla when group 0 server dies and for a good reason -- the data path may still work and killing Scylla would only hurt client applications.

In the same spirit we should not kill Scylla if topology coordinator fiber dies. It will only block cluster metadata operations but the data path may be functioning.

But I'm not sure if there is any reason (except shutdown and leader stepdown) for the topology coordinator fiber not to run forever

Reply:
If you remember there we complains already about our decision to not kill Scylla if group0 dies. When only schema changes worked through it it made some sense, but now it just makes a lot of thing non functional without easy way to understand why.

Reply:
> If you remember there we complains already about our decision to not kill Scylla if group0 dies.

I don't remember complaints, I remember questions. If there was a complaint then it came from a person who didn't understand why such decision was made, and the complaint stopped once we explained the reason.

> When only schema changes worked through it it made some sense, but now it just makes a lot of thing non functional without easy way to understand why.

If it's not easy to understand it's a documentation issue.

We should not needlessly disrupt client applications.
The choice is between:
- a bug kills the entire Scylla cluster, all apps stop working, millions of $$ loss, customer is furious, you get paged at night
- a bug kills topology coordinator / group 0 server on all nodes, but applications are still working. You sleep through the night and investigate the issue in the morning. No $$ loss

a simple `bad_alloc` is all it takes

The choice is obvious

Reply:
> a simple bad_alloc is all it takes

And I'm mentioning `bad_alloc` because we already saw them killing group 0 server. So it doesn't even need to be a "bug" per se

Reply:
What we can do to improve our testing is to call `on_internal_error` if group 0 server dies, for example. It won't cause production clusters to crash (unless client likes to play Russian roulette and uses `--abort-on-internal-error` which is not recommended, they asked for it) but it might make tests detect issues faster.

Reply:
> > If you remember there we complains already about our decision to not kill Scylla if group0 dies.
> 
> I don't remember complaints, I remember questions. If there was a complaint then it came from a person who didn't understand why such decision was made, and the complaint stopped once we explained the reason.
> 

Was from @tgrabiec iirc.

> > When only schema changes worked through it it made some sense, but now it just makes a lot of thing non functional without easy way to understand why.
> 
> If it's not easy to understand it's a documentation issue.

Documentation solves nothing here. Nobody is going to read a documentation at three o'clock to understand why a node cannot be replaced. And fixing a failed group0 meas you need to restart a node and I am sure admins will be happy to do it when they already missing one node.

> 
> We should not needlessly disrupt client applications. The choice is between:
> 
>     * a bug kills the entire Scylla cluster, all apps stop working, millions of $$ loss, customer is furious, you get paged at night
> 
>     * a bug kills topology coordinator / group 0 server on all nodes, but applications are still working. You sleep through the night and investigate the issue in the morning. No $$ loss
> 
> 
> a simple `bad_alloc` is all it takes

There are bad_alloc already that will kill a node. And bad_alloc usually do not kill entire cluster.

> 
> The choice is obvious

Not that obvious as you want it to be. Some problems better to fix as fast as possible and not wait for situation to aggravate. 

But I proposed to alternative. Another was to restart the fiber. Printing an error in the logs and continue as nothing happens is the worst of all the options.

Reply:
> Documentation solves nothing here. Nobody is going to read a documentation at three o'clock to understand why a node cannot be replaced.

Documentation consists not only of internet pages, but also on good log messages, good errors returned from APIs, good metrics and so on.

The choice is between:
- reading documentation at 3 o'clock while you can't replace a node 
- reading documentation at 3 o'clock while none of your applications are working at all because your whole DB crashed and you're losing $$ every second

> And fixing a failed group0 meas you need to restart a node and I am sure admins will be happy to do it when they already missing one node.

I'm sure admins will be even more happy if we do the shutdown part of that restart automatically for them, which is basically what you propose. And chance is on all nodes at the same time! (While they can do a rolling restart)

> And bad_alloc usually do not kill entire cluster.

Unless it's caused by applying some logic that is coincidentally being applied on every node due to a state machine replication mechanism

> Not that obvious as you want it to be. Some problems better to fix as fast as possible and not wait for situation to aggravate.

Group 0 server dying is a serious issue I agree, but that's why we have monitoring and **alerts**, and serious alerts from customers are already often a reason for Support getting called. In general if there's some serious trouble with node we already have a way to make everyone aware who needs to be, you don't need to make the situation even more serious than it is.

And we don't only print error in the logs, we also have the metric with group 0 status which should be used for triggering alerts (and if we don't do it, we should, @kostja was this ever communicated outside core team?)

Restarting the fiber is also an option, we can reach consensus there.

Reply:
> > > If you remember there we complains already about our decision to not kill Scylla if group0 dies.
> > 
> > 
> > I don't remember complaints, I remember questions. If there was a complaint then it came from a person who didn't understand why such decision was made, and the complaint stopped once we explained the reason.
> 
> Was from @tgrabiec iirc.
> 

Not sure which discussion you mean, maybe this one: https://github.com/scylladb/scylladb/issues/16493#issuecomment-1878762701
I was against aborting the node on failure to apply a command.

Reply:
> Restarting the fiber is also an option, we can reach consensus there.

Although we should probably still trigger an alert if the situation is not resolved (e.g. we see that we keep restarting the fiber multiple times within a small time period)

Reply:
> > > > If you remember there we complains already about our decision to not kill Scylla if group0 dies.
> > > 
> > > 
> > > I don't remember complaints, I remember questions. If there was a complaint then it came from a person who didn't understand why such decision was made, and the complaint stopped once we explained the reason.
> > 
> > 
> > Was from @tgrabiec iirc.
> 
> Not sure which discussion you mean, maybe this one: [#16493 (comment)](https://github.com/scylladb/scylladb/issues/16493#issuecomment-1878762701) I was against aborting the node on failure to apply a command.

It was not about group0 IIRC but topology coordinator fiber. In the end we made it noexcept.

Reply:
> We do not discuss group0 zero here. No need to attack a straw man you made here for yourself.

I'm not arguing with a strawman but with what you said:

> If you remember there we complains already about our decision to not kill Scylla if group0 dies. When only schema changes worked through it it made some sense, but now it just makes a lot of thing non functional without easy way to understand why.

That was about group 0 server, no?

Reply:
> > If you remember there we complains already about our decision to not kill Scylla if group0 dies. When only schema changes worked through it it made some sense, but now it just makes a lot of thing non functional without easy way to understand why.
> 
> That was about group 0 server, no?

Yes, but here we discuss a fiber that does not have any global effect, so why go on the tangent. 

With group0 there is a potential to have a bug that may kill all nodes, but even there there are cases where you can tell that it is just a node local out of memory. And nobody wants to micromanage cluster nodes. If a restart may improve the situation 10 out of 10 admins will chose restart instead of paging. Of course the best is to have an integration with a cluster management so that a node may ask the manager to be restarted and then if a manager gets to much of those requests it will do it one by one, but this is not something we have now.